Add container.lifecycle.type to api (sidecars KEP)

Joseph-Irving opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This implements the API for the Sidecar containers KEP https://github.com/kubernetes/enhancements/issues/753
It adds a new alpha optional field `type` to `container.lifecycle`, this is a string alias which can either be "Standard" or "Sidecar".
I will have follow up PRs based off of this to implement the logic in the Kubelet.

**Special notes for your reviewer**:
I've primarily based the implementation of this after ProcMount, which was the most recent example of an alpha string alias field that I could find.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
`Type` has been added to `container.Lifecycle` for the Sidecars alpha feature
```

/sig apps
/sig node

Joseph-Irving commented:
/label api-review

nilebox commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2689,6 +2711,20 @@ func validateContainers(containers []core.Container, isInitContainers bool, volu
+		}
+	}
+
+	return append(allErrors, field.Forbidden(fldPath, fmt.Sprint("must not only have containers with lifecyle type == sidecar in a pod")))
```
"must have at least one container with lifecycle type != sidecar"

nilebox replied:
or "... with a standard lifecycle"

Joseph-Irving replied:
yeah "must have at least one container with lifecycle type != sidecar" I think works better, specifying "with a standard lifecycle" may be confusing as this field is optional, so nil is fine

nilebox commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2689,6 +2711,20 @@ func validateContainers(containers []core.Container, isInitContainers bool, volu
+	allErrors := field.ErrorList{}
+
+	for _, container := range containers {
+		if container.Lifecycle == nil || container.Lifecycle.Type == nil || *container.Lifecycle.Type != core.LifecycleTypeSidecar {
```
Why don't we check for `container.Lifecycle.Type == core.LifecycleTypeStandard` instead?

Joseph-Irving replied:
yeah seems safer to do that, changed

nilebox reviewed (commented):
Great to finally see this feature being implemented, just left a couple nits.

draveness reviewed (commented):
/ok-to-test
/priority important-soon

Joseph-Irving commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2689,6 +2711,20 @@ func validateContainers(containers []core.Container, isInitContainers bool, volu
 }
 
+//check that we have at least one container in the pod which is not set to lifecyle type sidecar
+func validateSidecars(containers []core.Container, fldPath *field.Path) field.ErrorList {
```
agreed, changed to `validateHasNonSidecar`

Joseph-Irving commented:
thanks for the review @nilebox, I've hopefully addressed your comments.

I had to regenerate the test data to get the bazel-test to work, I did this by running `UPDATE_COMPATIBILITY_FIXTURE_DATA=true make test`, I couldn't find this documented anywhere, but appears to have fixed it, I put the new test data in second commit to make it more readable.

/test pull-kubernetes-integration

liggitt commented:
> I had to regenerate the test data to get the bazel-test to work, I did this by running `UPDATE_COMPATIBILITY_FIXTURE_DATA=true make test`, I couldn't find this documented anywhere, but appears to have fixed it

Doc in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/testdata/README.md#current-version (also, `make update` will do this along with other regeneration).

> I put the new test data in second commit to make it more readable.

Perfect, thanks

liggitt commented:
/assign @thockin @smarterclayton 
seemed to have the most context from https://github.com/kubernetes/enhancements/pull/919#pullrequestreview-244133225

rlenglet commented:
For this to work for Istio's sidecar proxies, we must make sure that Sidecar proxy containers are ready before the pod's application init containers start, in order for those init containers to have network connectivity. **So the `Sidecar` containers must be running before any init container starts.** 

So pod startup must look like:

1. sidecar init containers start
1. sidecar init containers finish
1. sidecar containers start
1. sidecar containers become ready
1. init containers start
1. init containers finish
1. containers start
1. containers become ready

From what I read in the [enhancement's doc](https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/sidecarcontainers.md), this doesn't seem to be the case in this proposal?!

Joseph-Irving commented:
@rlenglet I don't think this is the right place to have this discussion, I'd rather keep PR comments about what's in the PR. I think the enhancements issue https://github.com/kubernetes/enhancements/issues/753 or at sig-apps meeting/mailing-list are probably better places.

short answer: I don't think what you've described is required for this proposal to work with istio, I think you've described how to solve the problem of istio not working for init containers. We have discussed this, we have left the way open for something like init sidecars to be implemented in the future but it is not in scope for this KEP, there was already a fair amount of opposition to complicating the lifecycle this much, having containers that start before/during the init phase is a whole different discussion.

thockin commented:
Is there a summary of the discussion that we can share?  I may have been
part of it, but I don't remember it all :)

I just want to buy some modicum of confidence that we don't regret this
design later :)

On Mon, Jul 8, 2019 at 2:58 PM Joseph Irving <notifications@github.com>
wrote:

> @rlenglet <https://github.com/rlenglet> I don't think this is the right
> place to have this discussion, I'd rather keep PR comments about what's in
> the PR. I think the enhancements issue kubernetes/enhancements#753
> <https://github.com/kubernetes/enhancements/issues/753> or at sig-apps
> meeting/mailing-list are probably better places.
>
> short answer: I don't think what you've described is required for this
> proposal to work with istio, I think you've described how to solve the
> problem of istio not working for init containers. We have discussed this,
> we have left the way open for something like init sidecars to be
> implemented in the future but it is not in scope for this KEP, there was
> already a fair amount of opposition to complicating the lifecycle this
> much, having containers that start before/during the init phase is a whole
> different discussion.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/79649?email_source=notifications&email_token=ABKWAVBFI7FQRVNQDGXWCYTP6O2APA5CNFSM4H42RNKKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZOPBCY#issuecomment-509407371>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKWAVBQYRJSLSCGSCQCPFTP6O2APANCNFSM4H42RNKA>
> .
>

Joseph-Irving commented:
@thockin, I have tried my best to find any comments relating to init (you were indeed involved). Is it a bit tricky when split between two repos and a mixture of prs/issues etc. As well as some of this discussion taking place at various sig meetings.

Maybe I should write something up and add it to the KEP?

My TLDR version: 
- there are some use cases for having sidecars during the init phase (mainly istio)
- some sidecars require starting after init containers 
- a proposed solution is to allow `lifecycle.type: Sidecar` on init containers, these sidecar init containers would run alongside the init containers for the duration of the init phase only
- this wasn't considered to be needed for this initial alpha as it could be done later
- it will be harder to convince people to do this as it is a more significant change to pod/container lifecycle 

 
Links:
https://github.com/kubernetes/enhancements/pull/841#discussion_r257906512
https://github.com/kubernetes/community/pull/2148#discussion_r221094552
https://github.com/kubernetes/community/pull/2148#issuecomment-404415697
https://github.com/kubernetes/enhancements/issues/753#issuecomment-463681032

vllry commented:
/cc

Joseph-Irving committed:
add container.lifecycle.type to api

Joseph-Irving commented:
/test pull-kubernetes-bazel-test
/test pull-kubernetes-kubemark-e2e-gce-big

Joseph-Irving commented:
Hey @thockin @smarterclayton, I don't know if either of you have had the time to look at this yet, please let me know if there's anything I should be doing!

Joseph-Irving commented:
/milestone v1.17

zhan849 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2586,9 +2586,27 @@ func validateLifecycle(lifecycle *core.Lifecycle, fldPath *field.Path) field.Err
+	switch *lifecycleType {
+	case core.LifecycleTypeSidecar, core.LifecycleTypeStandard:
+		break
+	case "":
```
these 2 cases can be combined as you only defined "Standard" and "Sidecar" types

Joseph-Irving replied:
sorry I don't follow, can you explain what you mean?

thockin replied:
Yes - this is an optional field, there can't be a field.Required error

Joseph-Irving replied:
yeah, removed that case

zhan849 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2767,6 +2785,10 @@ func validateContainers(containers []core.Container, isInitContainers bool, volu
+		allErrs = append(allErrs, validateHasNonSidecar(containers, fldPath)...)
+	}
+
 	if isInitContainers {
```
since in alpha feature we have decided not to add sidecar container among init containers, shall we validate this as well?

Joseph-Irving replied:
Currently init containers don't support the Lifecycle field at all and they already have a validation for that. We could also add a check to make sure they don't have lifecycle type, not sure if necessary? or maybe a good idea to catch a regression?

thockin replied:
Validation already errors when lifecycle is set at all for init containers.

zhan849 reviewed (commented):
Thanks @Joseph-Irving this finally got kicked off!

zhan849 commented on pkg/apis/core/types.go:
```diff
@@ -2102,6 +2112,11 @@ type Lifecycle struct {
+	// One of Standard, Sidecar.
+	// Defaults to Standard.
+	// +optional
+	Type *LifecycleType
```
do we need to add lifecycle info to container status as well? this would provide more information to evaluate pod status.

lavalamp commented:
> So the Sidecar containers must be running before any init container starts.

Wow, as a user with workloads that currently have both init and sidecar containers, I would definitely not expect that, mostly because that's not how it happens right now...

thockin commented:
I ack @rlenglet concerns.  I don't know if the right design would be for sidecars to start before init or for there to be distinct init sidecars.  Other than mesh proxies, do we have other examples we can draw on?

thockin commented on pkg/api/pod/util.go:
```diff
@@ -391,6 +391,20 @@ func dropDisabledFields(
 		})
 	}
 
+	if !utilfeature.DefaultFeatureGate.Enabled(features.SidecarLifecycle) && !lifecycleTypeInUse(oldPodSpec) {
```
This is a bit different than other APIs, which explicitly nil-out the fields when gated.

Is there a reason not to do that here?  I get that it makes code a bit simpler to assume a value, but strictly speaking, the field more or less doesn't exist when gated.

In other words, is this PR adding a GA field with an alpha value or an alpha field?

Joseph-Irving replied:
no I think what you said makes sense. The field itself is alpha not the value so setting type to nil seems sensible. Will fix

thockin commented on pkg/apis/core/types.go:
```diff
@@ -2124,6 +2134,11 @@ type Lifecycle struct {
 	// +optional
 	PreStop *Handler
+	// Type describes the startup/shutdown behaviour of the container.
+	// One of Standard, Sidecar.
```
please quot "Standard" and "Sidecar" here.

What happens if a user sets this on an init container?

Joseph-Irving replied:
You'll get an API validation error as 'Lifecycle' is not supported on init containers. That aside nothing would currently happen. Do you think this should be mentioned here?

thockin replied:
yes

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2780,6 +2802,20 @@ func validateContainers(containers []core.Container, isInitContainers bool, volu
+	for _, container := range containers {
+		if container.Lifecycle == nil || container.Lifecycle.Type == nil || *container.Lifecycle.Type == core.LifecycleTypeStandard {
+			//if we find a container that is not a sidecar, return
+			return allErrors
```
return nil would be clearer?

Joseph-Irving replied:
yep changed

thockin reviewed (commented):
I reviewed for API.  We don't want to merge API changes without an implementation, so you can fix up the comments I made, and then rebase impl on this or just lump it together into a single PR.

zhan849 commented:
> I ack @rlenglet concerns. I don't know if the right design would be for sidecars to start before init or for there to be distinct init sidecars. Other than mesh proxies, do we have other examples we can draw on?

Hi @thockin thanks for checking out the sidecar code review. Regarding feature readiness, we (Pinterest) and Lyft has our own sidecar impl running in large scale production and we’ve summarize lessons we learned on implement this. See thread https://github.com/kubernetes/kubernetes/pull/80744#issuecomment-542299233 for your information about sidecar api / feature designs

Joseph-Irving commented:
thank you very much for the review @thockin, I've addressed your comments.

In regards to the actual implementation, that is currently located in this PR https://github.com/kubernetes/kubernetes/pull/80744 which builds off the two commits from this PR. I thought it would be easier to review if they were separate PRs (perhaps not).
So I can merge them together into one (or just close this one?), whatever is easiest for people.

Joseph-Irving committed:
regen test data

thockin commented on pkg/api/pod/util.go:
```diff
@@ -385,6 +385,19 @@ func dropDisabledFields(
 		})
 	}
 
+	if !utilfeature.DefaultFeatureGate.Enabled(features.SidecarLifecycle) && !lifecycleTypeInUse(oldPodSpec) {
```
I think you want to check this condition on each container and initContainer.  You don't want to allow new uses to this field on container B just because container A happened to use it.

It's really minor corner case, I admit

Joseph-Irving replied:
How would you be sure that the container in the new pod spec is the same as the container in the old pod spec? Both the name of the container and its position could be changed in an update. 

e.g
old pod
```yaml
containers:
- name: usesSidecar
  image: sidecar
  lifecycle: sidecar
- name: normalContainer
  image: myapp
```
could change to
new pod
```yaml
containers:
- name: normalContainer
  image: myapp
- name: usesSidecarNewName
  image: sidecar
  lifecycle: sidecar
```
In this case I'm not sure how I would verify that container `usesSidecarNewName` used the sidecar lifecycle in the previous spec, and would therefore remove it, losing info and changing behaviour unintentionally. Maybe I'm missing something obvious.

thockin replied:
Good point.  The list declares "name" as the key, so the correct-est answer would be to let it remain on any container with the same name, but not allow it on containers with new names.

If that's easy, great.  It may simply not be worthwhile, in which case just comment to explain why this shortcut is taken - basically, "yes, it's not exactly right, but doing it right is not worth the effort".

Thoughts?

Joseph-Irving replied:
ok thanks that was the clarification I needed, I have this got to work and added a test for it, so it will now check not if the Pod was using the lifecycle but if the container with the same name in the old pod spec was using it.

thockin commented on pkg/apis/core/types.go:
```diff
@@ -2124,6 +2134,11 @@ type Lifecycle struct {
 	PreStop *Handler
+	// Type describes the startup/shutdown behaviour of the container.
+	// One of "Standard", "Sidecar".
+	// Defaults to "Standard".
```
You say it default to Standard, but that is not implemented anywhere - there should be a corresponding change to pkg/apis/core/v1/defaults.go that explicitly sets the value, right?

Joseph-Irving replied:
I wasn't entirely sure about the convention on this. How much should we be defaulting for a new alpha field? Currently `Lifecycle` is not a default field on a container so should we be adding `Lifecycle` + `Lifecycle.Type` to all containers and setting that to `Standard`? Or should we only default `Lifecycle.Type` if `Lifecycle != nil`?

Joseph-Irving replied:
I also seem to recall some issues with defaulting PodSpec values as it would cause rollouts on a version upgrade, such as in https://github.com/kubernetes/kubernetes/pull/78885 and https://github.com/kubernetes/kubernetes/pull/69988

thockin replied:
@liggitt since you were active on #78885

What is the correct course for default fields that end up underneath pods?  It seems that the pattern from #69988 is to set if for Pod but not all PodSpec. (frankly, the type-based defaulting was/is a bad idea, we should fix that).

liggitt replied:
Defaults in podspecs are particularly twisty. The safest thing to do is to avoid any defaulting that would change values in existing pods. For example, a default value for a field like this would could change content for existing pods. A default value for a field inside a new stanza (e.g. a new volume source type) would not change existing pods, because no existing pods would have that stanza.

The three areas podspec defaults have bitten us are:

1. New fields in podspec defaulted inside a deployment/replicaset/etc making workload controllers trigger new rollouts on API server upgrade. From what I understand, this is due to fragile computation by workload controllers of when the embedded pod spec has "changed" (e.g. hashing the pod spec, which is confused when a new field with a default value appears). This test catches additions like that:

https://github.com/kubernetes/kubernetes/blob/66262e351dad61cd70bd1591c7e69a3e5d218a28/pkg/apis/core/v1/defaults_test.go#L43-L52

2. New fields in podspec defaulted inside a pod (especially inside a container) making nodes restart pods on API server upgrade, due to the way the node detects whether the container spec it was asked to run has changed. This test catches additions like that:

https://github.com/kubernetes/kubernetes/blob/66262e351dad61cd70bd1591c7e69a3e5d218a28/pkg/apis/core/v1/defaults_test.go#L164-L172

3. New fields appearing anywhere within the podspec schema on API server upgrade (whether the field had a value or not) used to make nodes restart pods because they used to hash the go schema; that was fixed in https://github.com/kubernetes/kubernetes/pull/57741 in 1.16

Joseph-Irving replied:
So in this scenario, if we were to default `Lifecycle.Type` that would cause changes to existing Pods as we would be changing the values in `Lifecycle`, hence causing restarts/rollouts. Therefore we shouldn't be doing any defaulting here to avoid that from happening. 
I can change the comments etc, so that there is no reference to `Standard` being the default, just that it can be `Sidecar` or `Standard`. There are no assumptions in the Kubelet code that non-sidecars will have `Standard` set on them, so it isn't an issue from that perspective.

Joseph-Irving replied:
@thockin, any thoughts on this? I believe this is the last point in your review so once we've decided how to proceed with this, hopefully we'll be done :)

thockin replied:
I hate the limitation.  Hashing all of podspec is a terrible idea.  That said, it's what happens.

If we do not actually set the value, can we can assert that unset means "normal".  That makes it awkward if we ever want to change the default in a future API version, but that seems quite unlikely.

If that's the path, and it sounds like that is what @liggitt is saying, then we should document that and remove the "Standard" value entirely.

Joseph-Irving replied:
ok thanks that clear things up, so I shall remove "Standard" as it's somewhat redundant, and document that unset means you're following the "normal" lifecycle as opposed to the sidecar one, the kubelet code already makes this assumption so this will be a simple change to make. Thanks for the quick response :)

Joseph-Irving replied:
Ok, I've updated this to remove the "standard" lifecycle type, if you're now happy with this PR, we can close this and move over to the implementation PR https://github.com/kubernetes/kubernetes/pull/80744

thockin reviewed (commented):
Two points then I am happy with the API review.

thockin commented:
FWIW, we won't merge this but the PR that has the impl *and* this

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2589,9 +2589,25 @@ func validateLifecycle(lifecycle *core.Lifecycle, fldPath *field.Path) field.Err
 
+var supportedLifecycleTypes = sets.NewString(string(core.LifecycleTypeSidecar))
+
+func validateLifecycleType(lifecycleType *core.LifecycleType, fldPath *field.Path) field.ErrorList {
```
Do we need to error is isInitContainer && has sidecar?  The API comments say that is an error

Joseph-Irving replied:
having the `Lifecycle` field is already a validation error for init containers
```go
	if ctr.Lifecycle != nil {
			allErrs = append(allErrs, field.Invalid(idxPath.Child("lifecycle"), ctr.Lifecycle, "must not be set for init containers"))
		}
```
so I wasn't sure if we needed the extra validation step for `Lifecycle.Type`

thockin replied:
ACK

thockin commented:
API is good except validation question.  I will shift to impl PR.

/lgtm
/hold

dims commented:
/milestone v1.18

dims commented:
added milestone as https://github.com/kubernetes/enhancements/issues/753 is marked 1.18

smourapina commented:
@Joseph-Irving, @thockin:
This is Bug Triage for release 1.18. I would like to leave a friendly reminder that code freeze is in a few days (Thursday 5 March), so please take this into account. Thanks!

dims commented:
/assign @thockin

Joseph-Irving commented:
api review has been completed and the commits from here are contained in this PR https://github.com/kubernetes/kubernetes/pull/80744, so I will close this one so that the api and implementation and api can be merged together
